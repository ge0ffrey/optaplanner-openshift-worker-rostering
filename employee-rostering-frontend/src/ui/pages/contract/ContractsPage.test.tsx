/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { ContractsPage, Props } from './ContractsPage';

describe('Contracts page', () => {
  it('should render correctly with no contracts', () => {
    const contractsPage = shallow(<ContractsPage {...noContracts} />);
    expect(toJson(contractsPage)).toMatchSnapshot();
  });

  it('should render correctly with a few contracts', () => {
    const contractsPage = shallow(<ContractsPage {...twoContracts} />);
    expect(toJson(contractsPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = twoContracts.tableData[1];
    const viewer = shallow(contractsPage.renderViewer(contract));
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = twoContracts.tableData[1];
    const editor = shallow(contractsPage.renderEditor(contract));
    expect(toJson(editor)).toMatchSnapshot();
  });

  it('should call addContract on addData', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = twoContracts.tableData[1];
    contractsPage.addData(contract);
    expect(twoContracts.addContract).toBeCalled();
    expect(twoContracts.addContract).toBeCalledWith(contract);
  });

  it('should call updateContract on updateData', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = twoContracts.tableData[1];
    contractsPage.updateData(contract);
    expect(twoContracts.updateContract).toBeCalled();
    expect(twoContracts.updateContract).toBeCalledWith(contract);
  });

  it('should call removeContract on removeData', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = twoContracts.tableData[1];
    contractsPage.removeData(contract);
    expect(twoContracts.removeContract).toBeCalled();
    expect(twoContracts.removeContract).toBeCalledWith(contract);
  });

  it('should treat empty name as invalid', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = {...twoContracts.tableData[1], name: ""};
    const result = contractsPage.isValid(contract);
    expect(result).toEqual(false);
  });

  it('should treat non-empty name as valid', () => {
    const contractsPage = new ContractsPage(twoContracts);
    const contract = {...twoContracts.tableData[1], name: "Contract"};
    const result = contractsPage.isValid(contract);
    expect(result).toEqual(true);
  });
});

const noContracts: Props = {
  tenantId: 0,
  title: "Contracts",
  columnTitles: ["Name", "Max Hours Per Day", "Max Hours Per Week", "Max Hours Per Month", "Max Hours Per Year"],
  tableData: [],
  addContract: jest.fn(),
  updateContract: jest.fn(),
  removeContract: jest.fn()
};

const twoContracts: Props = {
  tenantId: 0,
  title: "Contracts",
  columnTitles: ["Name", "Max Hours Per Day", "Max Hours Per Week", "Max Hours Per Month", "Max Hours Per Year"],
  tableData: [{
    id: 0,
    version: 0,
    tenantId: 0,
    name: "Contract 1",
    maximumMinutesPerDay: null,
    maximumMinutesPerWeek: null,
    maximumMinutesPerMonth: null,
    maximumMinutesPerYear: null
  },
  {
    id: 1,
    version: 0,
    tenantId: 0,
    name: "Contract 2",
    maximumMinutesPerDay: 1,
    maximumMinutesPerWeek: null,
    maximumMinutesPerMonth: 10,
    maximumMinutesPerYear: 120
  }],
  addContract: jest.fn(),
  updateContract: jest.fn(),
  removeContract: jest.fn()
};