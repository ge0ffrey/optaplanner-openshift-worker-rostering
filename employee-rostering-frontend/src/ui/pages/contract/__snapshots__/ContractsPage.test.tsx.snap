// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contracts page should render correctly with a few contracts 1`] = `
<div>
  <Button
    isDisabled={false}
    onClick={[Function]}
  >
    Add
  </Button>
  <Table
    borders={true}
    caption="Contracts"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Day",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Week",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Month",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Year",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onCollapse={null}
    onExpand={null}
    role="grid"
    rowLabeledBy="simple-node"
    rows={
      Array [
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Contract 1"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                viewer={
                  <Text>
                    Contract 1
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={null}
                    isValid={[Function]}
                    label="Max minutes per day"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={<Text />}
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={null}
                    isValid={[Function]}
                    label="Max minutes per week"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={<Text />}
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={null}
                    isValid={[Function]}
                    label="Max minutes per month"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={<Text />}
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={null}
                    isValid={[Function]}
                    label="Max minutes per year"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={<Text />}
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span>
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                viewer={
                  <span>
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Contract 2"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                viewer={
                  <Text>
                    Contract 2
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={1}
                    isValid={[Function]}
                    label="Max minutes per day"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={
                  <Text>
                    1
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={null}
                    isValid={[Function]}
                    label="Max minutes per week"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={<Text />}
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={10}
                    isValid={[Function]}
                    label="Max minutes per month"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={
                  <Text>
                    10
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <OptionalInput
                    defaultValue={120}
                    isValid={[Function]}
                    label="Max minutes per year"
                    onChange={[Function]}
                    valueMapper={[Function]}
                    valueToString={[Function]}
                  />
                }
                viewer={
                  <Text>
                    120
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span>
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                viewer={
                  <span>
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
      ]
    }
    variant={null}
  >
    <TableHeader
      className=""
    />
    <TableBody
      className=""
      onRowClick={[Function]}
      rowKey="id"
    />
  </Table>
</div>
`;

exports[`Contracts page should render correctly with no contracts 1`] = `
<div>
  <Button
    isDisabled={false}
    onClick={[Function]}
  >
    Add
  </Button>
  <Table
    borders={true}
    caption="Contracts"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Day",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Week",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Month",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Max Hours Per Year",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onCollapse={null}
    onExpand={null}
    role="grid"
    rowLabeledBy="simple-node"
    rows={Array []}
    variant={null}
  >
    <TableHeader
      className=""
    />
    <TableBody
      className=""
      onRowClick={[Function]}
      rowKey="id"
    />
  </Table>
</div>
`;

exports[`Contracts page should render the editor correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <TextInput
      aria-label="Name"
      className=""
      defaultValue="Contract 2"
      isDisabled={false}
      isReadOnly={false}
      isRequired={false}
      isValid={true}
      key="0"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </td>
  <td
    key="1"
  >
    <OptionalInput
      defaultValue={1}
      isValid={[Function]}
      key="1"
      label="Max minutes per day"
      onChange={[Function]}
      valueMapper={[Function]}
      valueToString={[Function]}
    />
  </td>
  <td
    key="2"
  >
    <OptionalInput
      defaultValue={null}
      isValid={[Function]}
      key="2"
      label="Max minutes per week"
      onChange={[Function]}
      valueMapper={[Function]}
      valueToString={[Function]}
    />
  </td>
  <td
    key="3"
  >
    <OptionalInput
      defaultValue={10}
      isValid={[Function]}
      key="3"
      label="Max minutes per month"
      onChange={[Function]}
      valueMapper={[Function]}
      valueToString={[Function]}
    />
  </td>
  <td
    key="4"
  >
    <OptionalInput
      defaultValue={120}
      isValid={[Function]}
      key="4"
      label="Max minutes per year"
      onChange={[Function]}
      valueMapper={[Function]}
      valueToString={[Function]}
    />
  </td>
</tr>
`;

exports[`Contracts page should render the viewer correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <Text
      key="0"
    >
      Contract 2
    </Text>
  </td>
  <td
    key="1"
  >
    <Text
      key="1"
    >
      1
    </Text>
  </td>
  <td
    key="2"
  >
    <Text
      key="2"
    />
  </td>
  <td
    key="3"
  >
    <Text
      key="3"
    >
      10
    </Text>
  </td>
  <td
    key="4"
  >
    <Text
      key="4"
    >
      120
    </Text>
  </td>
</tr>
`;
